core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -204,13 +204,20 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
           mod1 >= mod2
       }
 
-      logInfos.sliding(20, 20).foreach { batch =>
+      val tasks: Iterator[Future[_]] = logInfos.grouped(20).map { batch =>
```

Comment:
Since we're talking about cleaning this up, you could do this also:

    logInfos.grouped(20)
      .map { batch =>
        replayExecutor.submit(new Runnable {
          override def run(): Unit = mergeApplicationListing(batch)
        })
      }
      .foreach { task =>
        // Wait for all tasks to finish. This makes sure that checkForLogs is
        // not scheduled again while some tasks are already running in the
        // replayExecutor.
        try {
          task.get()
        } catch {
          case e: InterruptedException =>
            throw e
          case e: Exception =>
            logWarning("Error replaying logs.", e)
        }
      }

Note I added some missing exception handling, which would cause you to revert to the old behavior of piling up executions if an error happened.